File: Retos/Reto #3 - EL GENERADOR DE CONTRASEÑAS [Media]/c/Old-Lx.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>

int RandomInt(int upper, int defSeed) {
    ///Inspirado por https://www.geeksforgeeks.org/generating-random-number-range-c/
    srand(time(NULL) * upper + defSeed);
    int num = (rand() % (upper + 1));
    return num;
}

int setMayus() {
    
    char m[1];
    printf("¿Deseas que tu contraseña tenga mayusculas?\n[1]Si\n[0]No\n");
    fflush(stdout);
    scanf("%s", m);
    fflush(stdin);
    int mayus = atoi(m);
    switch (mayus) {
        case 0:
            return 0;
            break;

        case 1:
            return 1;
            break;
        
        default:
            printf("Error debe seleccionar [1]Si o [0]No\n\n");
            setMayus();
            break;
    }
}

int setNum() {
    
    char n[1];
    printf("¿Deseas que tu contraseña tenga numeros?\n[1]Si\n[0]No\n");
    fflush(stdout);
    scanf("%s", n);
    fflush(stdin);
    int num = atoi(n);
    switch (num) {
        case 0:
            return 0;
            break;

        case 1:
            return 1;
            break;
        
        default:
            printf("Error debe seleccionar [1]Si o [0]No\n\n");
            setNum();
            break;
    }
}

int setSym() {
    
    char s[1];
    printf("¿Deseas que tu contraseña tenga simbolos?\n[1]Si\n[0]No\n");
    fflush(stdout);
    scanf("%s", s);
    fflush(stdin);
    int sym = atoi(s);
    switch (sym) {
        case 0:
            return 0;
            break;

        case 1:
            return 1;
            break;
        
        default:
            printf("Error debe seleccionar [1]Si o [0]No\n\n");
            setSym();
            break;
    }
}

int setTam() {
    
    char tama[2];
    printf("Indique la longitud de su contraseña (entre 8 y 16 caracteres):\n");
    fflush(stdout);
    scanf("%s", tama);
    fflush(stdin);
    int tam = atoi(tama);
    return tam;
}

int setGenLength(int mayus, int num, int sym) {
    
    int genLength;

    if (mayus) {
        genLength = 52;
    } else {
        genLength = 26;
    }

    if (num) {
        genLength = genLength + 10;
    }

    if (sym) {
        genLength = genLength + 7;
    }

    return genLength;
}

char *setGen(int mayus, int num, int sym) {
    
    char *minusLetters = "abcdefghijklmnopqrstuvwxyz";
    char *letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    char *numbers = "0123456789";
    char *symbols = ",.-#'?!"; 
    int length = setGenLength(mayus, num, sym);
    char *generator = (char *)malloc(sizeof(char) * length);
    if (mayus) {
        strcpy(generator, letters);
    } else {
        strcpy(generator, minusLetters);
    }

    if (num) {
        strcat(generator, numbers);
    }

    if (sym) {
        strcat(generator, symbols);
    }

    return generator;
}

int strLength(char *string) {
    
    int i = 0;
    while (string[i] != '\0') {
        i++;
    }
    return i;
}

char *passwordGen(int tam) {
    
    int defSeed = 73489438;

    if (tam < 8 || tam > 16) {
        printf("Debe seleccionar entre 8 y 16 caracteres\n\n");
        passwordGen(setTam());
    }

    char *password = (char *)malloc(sizeof(char) * tam + 1);
    char *generator = setGen(setMayus(), setNum(), setSym());

    int length = strLength(generator);
    
    for (int i = 0; i < tam; i++) {
        defSeed = RandomInt(length - 1, defSeed);
        password[i] = generator[RandomInt(length - 1, defSeed)];
    }
    password[tam] = '\0';

    printf("Tu contraseña es: %s\n", password);

    char ans[1];
    printf("¿Deseas generar otra?\n[1]Si\n[0]No\n");
    fflush(stdout);
    scanf("%s", ans);
    fflush(stdout);
    int repeat = atoi(ans);

    switch (repeat) {
    case 0:
        break;
    
    case 1:
        passwordGen(setTam());
    
    default:
        break;
    }

    return password;
    
}

int main() {
    
    passwordGen(setTam());
}

File: Retos/Reto #3 - EL GENERADOR DE CONTRASEÑAS [Media]/python/Old-Lx.py
from random import choice
from string import ascii_letters
from string import ascii_lowercase
from string import digits
from string import punctuation

def set_values():
    print("Indique la cantidad de caracteres (debe estar entre 8 y 16):")
    length = int(input())
    print("¿Desea que su contraseña incluya mayúsculas?\n[1]Sí\n[0]No\n")
    mayus = int(input())
    print("¿Desea que su contraseña incluya números?\n[1]Sí\n[0]No\n")
    numbers = int(input())
    print("¿Desea que su contraseña incluya símbolos?\n[1]Sí\n[0]No\n")
    symbols = int(input())
    print("\n")
    values = [length, mayus, numbers, symbols]
    return values

def gen_password(length: int, mayus: bool = True, numbers: bool = True, symbols: bool = True):
    
    if length < 8 or length > 16:
        print("La contraseña debe ser de 8 a 16 caracteres\n")
        values = set_values()
        gen_password(values[0], values[1], values[2], values[3])
    
    password = ""
    letters = ascii_letters
    minus_letters = ascii_lowercase
    num = digits
    sym = punctuation

    if mayus == True:
        generator = letters  
    else:
        generator = minus_letters

    if numbers == True:
        generator = generator + num
    
    if symbols == True:
        generator = generator + sym
    
    for i in range(length):
            password = password + choice(generator)
    print("Su contraseña es: [ " + password + " ]")
    return password

values = set_values()
gen_password(values[0], values[1], values[2], values[3])
